### PR TITLE
Autoscaler hotfix for #4555.

### PR DESCRIPTION
**Pull request by romilbhardwaj**
## What do these changes do?

These changes may fix the autoscaler bug introduced in #4555:

> With a head node with num-cpus=1, this commit results in the autoscaler LoadMetrics reporting an empty dict for both static and dynamic resouces once 1 job is scheduled on the head node.`

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by romilbhardwaj** [comment]
@ls-daniel Can you verify if this fixes the autoscaler? Or would it be possible to share the autoscaler config that was failing?

**Comment by jovany-wang** [comment]
Is this PR also able to fix https://github.com/ray-project/ray/issues/4655?

**Comment by ls-daniel** [comment]
This doesn't work.

The problem is that static resources also disappear.

If we add `print("static_resources", static_resources)` to that function, and start running some tasks, the logs look like this:


```2019-04-18 13:55:00,688 DEBUG static_resources {b'CPU': 1.0}
2019-04-18 13:55:00,805 DEBUG static_resources {b'CPU': 1.0}
2019-04-18 13:55:00,921 DEBUG static_resources {b'CPU': 1.0}
2019-04-18 13:55:01,030 DEBUG static_resources {}
2019-04-18 13:55:01,142 DEBUG static_resources {}
2019-04-18 13:55:01,253 DEBUG static_resources {}
```

that is, the static resources disappear.

**Comment by ls-daniel** [comment]
If I revert the other commit it works fine. I'm fairly sure it's not an issue in the autoscaler. I've had a look at the commit to try and figure out which bit of the code is adjusting static resources.

It makes sense to me that you may wish for a dynamic resource to disappear instead of being zero. A static resource should a) never be zero and b) never disappear, it's static :)

**Comment by romilbhardwaj** [comment]
Thanks @ls-daniel. I agree, static resources should indeed never be zero or disappear.

The `LoadMetrics.update` method is called from `monitor.py`, which updates it with values from the raylet heartbeats: 

https://github.com/ray-project/ray/blob/be2cbdf1306182007b904b5d976c0c318e6864e9/python/ray/monitor.py#L118-L131

This behavior is a bit unexpected since the `resources_total_` in `cluster_resource_map` is never modified once created from the GCS entry, and thus `ResourcesTotalCapacity` in the heartbeat should never change. cc @robertnishihara

**Comment by romilbhardwaj** [comment]
> Is this PR also able to fix #4655?

Yes, I think this is related to #4655 and should have fixed it.

File: python/ray/autoscaler/autoscaler.py [hunk 1]
```diff
@@ -158,6 +158,13 @@ def __init__(self):
...
     def update(self, ip, static_resources, dynamic_resources):
         self.static_resources_by_ip[ip] = static_resources
+        # TODO(romilb): This is a hotfix to make autoscaler work with #4555
+        dynamic_resources_update = dynamic_resources.copy()
+        for resource_name, capacity in static_resources.items():
+            if resource_name not in dynamic_resources_update:
+                dynamic_resources_update[resource_name] = 0.0
+        self.dynamic_resources_by_ip[ip] = dynamic_resources_update
+
         self.dynamic_resources_by_ip[ip] = dynamic_resources
```
Comment (dulex123) [inline python/ray/autoscaler/autoscaler.py]: Line 168 and Line 166 do not make sense together, should 168 be removed?

Comment (robertnishihara) [reply python/ray/autoscaler/autoscaler.py]: Yes I think so, thanks!

**Comment by nikola-j** [comment]
This change solved #4655 for me.

**Comment by ls-daniel** [comment]
I've tried with @rnishihara's change and that doesn't work either. Same behaviour.

The autoscaler on Ray master is broken for me, further changes to the resource mechanisms have meant reverting 4555 isn't easily possible any more either, so I'm stuck about 30 commits back now

**Comment by robertnishihara** [comment]
@ls-daniel thanks, @williamma12 is looking into it.

**Comment by williamma12** [comment]
@ls-daniel I was able to reproduce the error and pushed a fix that fixed it for me, can you let me know if it fixes it for you?

**Comment by ls-daniel** [comment]
@williamma12 this works perfectly cherry-picked into my branch - Thank you :)

I'm now rebasing on master to test again and will report back.

**Review by robertnishihara — APPROVED** [review]

**Comment by robertnishihara** [comment]
Thanks @williamma12! @ls-daniel The autoscaler is unfortunately not as well tested because it interacts with AWS or GCP. Do you have thoughts about how to improve the testing setup?

**Pull request merged**